File: BOJ/위상정렬/1766-문제집.java
import java.util.*;
import java.io.*;

class Main {
    static ArrayList<Integer>[] graph;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        graph = new ArrayList[n + 1];
        int[] inDegree = new int[n + 1];

        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            graph[x].add(y);
            inDegree[y]++;
        }

        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int i = 1; i <= n; i++) {
            if (inDegree[i] == 0) {
                pq.add(i);
            }
        }

        StringBuilder sb = new StringBuilder();

        while (!pq.isEmpty()) {
            int cur = pq.poll();
            sb.append(cur).append(" ");

            for (int next : graph[cur]) {
                inDegree[next]--;
                if (inDegree[next] == 0) {
                    pq.add(next);
                }
            }
        }

        System.out.println(sb.toString().trim());
    }
}

File: BOJ/위상정렬/1766-문제집.py
import sys, heapq  # 1번 문제가 가장 쉬운 문제 -> 우선순위큐 써야함

input = sys.stdin.readline

def bfs(n, graph, inDegree):
    answer = ""
    pq = []
    for i in range(1, n + 1):
        if inDegree[i] == 0:
            heapq.heappush(pq, i)

    while pq:
        cur = heapq.heappop(pq)
        answer += (str(cur) + " ")
        for next in graph[cur]:
            inDegree[next] -= 1 # cur가 실행되었으므로 next의 선행조건이 줄음
            if inDegree[next] == 0:
                heapq.heappush(pq, next)
    return answer


n, m = map(int, input().split())

graph = [[] for _ in range(n + 1)]
inDegree = [0] * (n + 1)

for i in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    inDegree[b] += 1

print(bfs(n, graph, inDegree))